File: partition-labels/partition-labels.java
class Solution {
    public List<Integer> partitionLabels(String s) {
        
        int lastIndex[] = new int[26];
        
        ArrayList<Integer> partitions = new ArrayList<>();
        
        // record the rightmost index of each letter in s
        for (int i = 0; i < s.length(); ++i) {
            
            lastIndex[s.charAt(i) - 'a'] = i;
        }
        
        // using the record of the rightmost indices as end points,
        // create partitions that contain every appearnce of the letters they hold
        for (int left = 0; left < s.length(); ++left) {
            
            int right = left;
            
            int partitionEnd = lastIndex[s.charAt(left) - 'a'];
            
            while (left < partitionEnd) {
                
                ++left;
                
                partitionEnd = Math.max(partitionEnd, lastIndex[s.charAt(left) - 'a']);
            }
            
            partitions.add(left - right + 1);
        }
        
        return partitions;
    }
}

File: partition-labels/partition-labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        
        int lastIndex[26] {0};
        
        vector<int> partitions {};
        
        // record the rightmost index of each letter in s
        for (int i = 0; i < s.size(); ++i) {
            
            lastIndex[s[i] - 'a'] = i;
        }
        
        // using the record of the rightmost indices as end points,
        // create partitions that contain every appearnce of the letters they hold
        for (int left = 0; left < s.size(); ++left) {
            
            int right = left;
            
            int partitionEnd = lastIndex[s[left] - 'a'];
            
            while (left < partitionEnd) {
                
                ++left;
                
                partitionEnd = max(partitionEnd, lastIndex[s[left] - 'a']);
            }
            
            partitions.push_back(left - right + 1);
        }
        
        return partitions;
    }
};